### PR TITLE
Feature/list published pagecontent if versioning enabled

### PR DESCRIPTION
**Pull request by NarenderRajuB**
Add filter to list only published page contents if versioning is enabled

### CODE REVIEW

File: djangocms_page_sitemap/sitemap.py [hunk 1]
```diff
@@ -54,6 +54,15 @@ def items(self):
 
             if page_url.language in excluded:
                 continue
+            if is_versioning_enabled():
+                # render only published version of page contents.
+                page_content = PageContent.objects.filter(
```
Comment (Aiky30) [inline djangocms_page_sitemap/sitemap.py:59]: This is really bad for performance. I think we need to query better in the first place.

File: djangocms_page_sitemap/sitemap.py [hunk 2]
```diff
@@ -31,14 +31,13 @@ def items(self):
             .filter(language__in=languages, path__isnull=False, page__login_required=False)
```
Comment (Aiky30) [inline djangocms_page_sitemap/sitemap.py:31]: We need to prefetch page__page_content, then in the url loop below (for page_url in all_urls:) we can exclude any urls that don't have a page_content entry.

File: djangocms_page_sitemap/sitemap.py [hunk 3]
```diff
@@ -54,6 +53,14 @@ def items(self):
```
Comment (Aiky30) [inline djangocms_page_sitemap/sitemap.py:53]: Maybe something like: if not page_url.page.page_content_count: continue

File: tests/test_sitemap.py [hunk 4]
```diff
@@ -17,22 +18,59 @@
 class SitemapTest(BaseTest):
 
     def test_sitemap_base(self):
-        test_string = '<url><loc>http://example.com/it/</loc><lastmod>%s</lastmod><changefreq>monthly</changefreq><priority>0.5</priority></url>' % now().strftime('%Y-%m-%d')
-        self.get_pages()
+        page1, page2, page3 = self.get_pages()
+        if is_versioning_enabled():
+            page1_content = create_title(title='pagecontent1', language='en', page=page1,
```
Comment (Aiky30) [inline tests/test_sitemap.py:23]: get_pages already creates the page contents. I would advise doing the publishing in one place i.e. get_pages, this is how it used ot work: https://github.com/nephila/djangocms-page-sitemap/blob/develop/tests/base.py#L37

**Review by Aiky30 — CHANGES_REQUESTED** [review]

File: djangocms_page_sitemap/sitemap.py [hunk 5]
```diff
@@ -23,22 +24,32 @@ def items(self):
         # https://github.com/divio/django-cms/blob/2894ae8bcf92092d947a097499c01ab2bbb0e6df/cms/sitemaps/cms_sitemap.py
         site = get_current_site()
         languages = get_public_languages(site_id=site.pk)
+        page_content_queryset = PageContent.objects.filter(
+            page__node__site=site,
+            language__in=languages
```
Comment (Aiky30) [inline djangocms_page_sitemap/sitemap.py:29]: I'm thinking that we can add  `page__pagesitemapproperties__include_in_sitemap=True` here now as we are using the prefetch query.

We then no longer need the exclusion filtering below.

We would need to do a count and prove this logic.

File: djangocms_page_sitemap/sitemap.py [hunk 6]
```diff
@@ -23,22 +24,32 @@ def items(self):
...
+        )
+        page_content_prefetch= Prefetch(
```
Comment (Aiky30) [inline djangocms_page_sitemap/sitemap.py:31]:
```suggestion
        page_content_prefetch = Prefetch(
```

**Review by Aiky30 — CHANGES_REQUESTED** [review]

File: djangocms_page_sitemap/sitemap.py [hunk 7]
```diff
@@ -54,6 +65,15 @@ def items(self):
 
             if page_url.language in excluded:
                 continue
+
+            published_page_content = False
```
Comment (Aiky30) [inline djangocms_page_sitemap/sitemap.py:69]:
```python
            has_page_content = False
            for page_content in page_url.page.pagecontent_set.all():
                if page_url.language == page_content.language:
                    has_page_content = True
            if has_page_content:
                valid_urls.append(page_url)
```

**Review by Aiky30 — APPROVED** [review]

**Comment by Aiky30** [comment]
@yakky We found that we were outputting draft pages in the sitemap, there appeared to be a large performance hit in the previous implementation of fetching the items, both are fixed with the changes in this PR.

**Review by yakky — APPROVED** [review]

**Pull request merged**